Move injectors to ManagedCredentialType

chrismeyersfsu opened the pull request:
Injectors are a bail-out mechanism for when you need to do something that the templating engine does not support, for a particular credential type. They are strongly tied to one and only one credential type. Before this change they lived far away from the ManagedCredentialType they are associated for. The link was the credential kind string was the same as the injector function name. This was a very loose coupling. This change is a tighter coupling.

chrismeyersfsu committed:
Move injectors to ManagedCredentialType

Injectors are a bail-out mechanism for when you need to do something
that the templating engine does not support, for a particular credential
type. They are strongly tied to one and only one credential type. Before
this change they lived far away from the ManagedCredentialType they are
associated for. The link was the credential kind string was the same as
the injector function name. This was a very loose coupling. This change
is a tighter coupling.

AlanCoding reviewed (approved):
Putting this in my own words...

The custom logic like `awx_plugins.credentials.injectors.aws` already lived here, but there was no link in the plugin definition to the custom logic. So somehow that got referenced by a literal name like `"aws"`.

My only gripe would be that `post_injectors` doesn't feel like a _great_ name. Like, sure, maybe it runs after the other stuff (templates)? But it's usually an _alternative_ to the templates, making that detail kind of irrelevant practically. It's more important that it runs custom python logic like a caveman, as opposed to any structured injector definition.

chrismeyersfsu commented:
1. `injectors_override`
2. `custom_injectors`
3. `alt_injectors`
4. `injectors_fn`

AlanCoding commented:
I'd lean towards (2) myself.

webknjaz commented:
I had an alternative idea with https://github.com/ansible/awx_plugins.interfaces/pull/10 / https://github.com/ansible/awx-plugins/pull/50 / https://github.com/ansible/awx/pull/15595 FTR

chrismeyersfsu committed:
Rename injector callback

* Defining an injector as code is an override. The previous name made it
  sound like the template engine would run, then the custom injector.
  This is not the case. If the custom injector is defined then the
  templating engine doens't run.

chrismeyersfsu committed:
Fix lint

AlanCoding commented on src/awx_plugins/credentials/plugins.py:
```diff
@@ -711,6 +726,7 @@
     namespace='kubernetes_bearer_token',
     kind='kubernetes',
     name=gettext_noop('OpenShift or Kubernetes API Bearer Token'),
+    injector=kubernetes_bearer_token,
```
```suggestion
    custom_injector=kubernetes_bearer_token,
```

AlanCoding commented:
I slightly lean towards the sentiment @chrismeyersfsu expressed in https://github.com/ansible/awx/pull/15595

> So I lean more towards getting rid of this "problem" by bringing these custom injectors closer to the ManagedCredentialType they are related to.

But even if we did nothing, it wouldn't bother me that much, since they already live in the plugins repo.

webknjaz commented:
@chrismeyersfsu typing change handling instructions:
1. ~Merge https://github.com/ansible/awx_plugins.interfaces/pull/13~
2. ~Click `Run workflow` @ https://github.com/ansible/awx_plugins.interfaces/actions/workflows/ci-cd.yml, entering `0.0.1a2` precisely in the version field. Don't enter anything else and submit the form.~
3. ~Wait until the newly created workflow run completes tests and pauses, then approve releasing on the workflow run summary page.~
4. ~Wait until the workflow run completes fully.~
5. Click `Run workflow` @ https://github.com/ansible/awx-plugins/actions/workflows/pip-tools.yml, entering `awx_plugins.interfaces` in the only input field.
6. Wait until that workflow completes and opens a PR with the version bump in the lock files.
7. Go to that PR and undraft it. You can approve it and enable auto-merge at this point.
8. Once that PR is merged, you can rebase this one.

webknjaz commented on src/awx_plugins/credentials/injectors.py:
```diff
@@ -6,6 +6,9 @@
 import stat
 import tempfile
 
+from awx_plugins.interfaces._temporary_private_api import (  # noqa: WPS436
```
```suggestion
from awx_plugins.interfaces._temporary_private_credential_api import (  # noqa: WPS436
```

webknjaz commented on src/awx_plugins/credentials/injectors.py:
```diff
@@ -16,7 +19,7 @@
 
 
-def aws(cred, env, private_data_dir):
+def aws(cred: Credential, env: dict, private_data_dir: str):
```
While on it, could you also mark the returned type for all of the functions in the module?
```suggestion
def aws(cred: Credential, env: dict, private_data_dir: str) -> None:
```

webknjaz commented on src/awx_plugins/credentials/plugins.py:
```diff
@@ -8,6 +8,16 @@
     gettext_noop,
 )
 
+from awx_plugins.credentials.injectors import (
```
Always use relative imports for local modules:
```suggestion
from .injectors import (
```

chrismeyersfsu committed:
Use new Credential type

* ManagedCredentialType custom callback functions now use a proper
  Credential class defined in awx_plugins.interfaces

chrismeyersfsu committed:
Tell docs to ignore not found Credential class

* I am not sure why the docs can not find the Credential class. Feels
  like maybe I'm pointing at a stale awx_plugins.interfaces. Also kind
  of feels like sphinx is confused by this repo being named awx_plugins
  and there being another repo/package with a similarish name so it is
  searching ansible/awx_plugins instead of
  ansible/awx_plugins.interfaces

webknjaz commented on src/awx_plugins/credentials/injectors.py:
```diff
@@ -9,14 +9,22 @@
 
 
-def aws(cred, env, private_data_dir):
+def aws(
+    cred: Credential,
+    env: dict[str, GenericOptionalPrimitiveType],
+    private_data_dir: str,
+) -> None:
```
FWIW, I recommend reverting the hack in #64 and adding docstrings to all of these instead.

webknjaz commented on .pre-commit-config.yaml:
```diff
@@ -194,7 +194,7 @@ repos:
     additional_dependencies:
     - >-  # standard interface declarations for AWX plugins
       awx_plugins.interfaces
-      @ git+https://github.com/ansible/awx_plugins.interfaces.git@ad4a965
+      @ git+https://github.com/ansible/awx_plugins.interfaces.git@ca7802a
```
I mentioned it somewhere already — best to drop Git references, as they are last resort.

```suggestion
    - awx_plugins.interfaces == 0.0.1a2
```

webknjaz replied:
fixed via #65

webknjaz commented on docs/conf.py:
```diff
@@ -209,4 +209,8 @@
 # Ref: https://stackoverflow.com/a/30624034/595220
 nitpick_ignore = [
     # temporarily listed ('role', 'reference') pairs that Sphinx cannot resolve
+    (
+        'py:class',
+        'awx_plugins.interfaces._temporary_private_credential_api.Credential',
+    ),
```
Recommended workaround: https://github.com/ansible/awx-plugins/pull/24/files#diff-85933aa74a2d66c3e4dcdf7a9ad8397f5a7971080d34ef1108296a7c6b69e7e3R209-R245

chrismeyersfsu committed:
Reduce diff coverage requirements
